src/webots/nodes/utils/WbPaintTexture.cpp

```diff
@@ -237,25 +237,27 @@ void WbPaintTexture::paint(const WbRay &ray, float leadSize, const WbRgb &color,
   if (mEvaporation && !gEvaporationTextures.contains(this))
     gEvaporationTextures.append(this);
 
-  const float blue = density * color.blue();
-  const float green = density * color.green();
-  const float red = density * color.red();
-  const float existingColorPercentage = 1.0f - density;
-
   const int halfSize2 = (halfSize * halfSize);
   for (int ty = oy; ty <= sy; ++ty) {
     int rowOffset = ty * w;
     int dataIndex = (rowOffset + ox) << 2;
     int circleCondition = halfSize2 - (ty - y) * (ty - y);
     for (int tx = ox; tx <= sx; ++tx) {
       if ((tx - x) * (tx - x) <= circleCondition) {
-        mData[dataIndex] = existingColorPercentage * mData[dataIndex] + blue;
-        mData[dataIndex + 1] = existingColorPercentage * mData[dataIndex + 1] + green;
-        mData[dataIndex + 2] = existingColorPercentage * mData[dataIndex + 2] + red;
-        mData[dataIndex + 3] = 1.0f;
-
-        if (mEvaporation)
-          mEvaporation[rowOffset + tx] = 1.0f;
+        const float previousDensity = mData[dataIndex + 3];
+        const float oldDensityRatio = previousDensity / (density + previousDensity);
+        mData[dataIndex] = oldDensityRatio * mData[dataIndex] + (1 - oldDensityRatio) * color.blue();
```

Comment:
I would use 1.0f also here
```suggestion
        mData[dataIndex] = oldDensityRatio * mData[dataIndex] + (1.0f - oldDensityRatio) * color.blue();
```